sdk/Trace/TracerProvider.php

```diff
@@ -6,69 +6,44 @@
 
 use OpenTelemetry\Trace as API;
 
-class TracerProvider implements API\TracerProvider
+final class TracerProvider implements API\TracerProvider
 {
-    /**
-     * @var self
-     */
-    protected static $instance;
 
     /**
      * @var Tracer[]
      */
     protected $tracers;
 
     /**
-     * @var SpanProcessor[]
+     * @var MultiSpanProcessor
      */
     protected $spanProcessors;
 
-    /**
-     * TracerFactory constructor.
-     *
-     * @param SpanProcessor[] $spanProcessors
-     */
-    final private function __construct(array $spanProcessors = [])
+    public function __construct()
     {
-        foreach ($spanProcessors as $spanProcessor) {
-            if (!$spanProcessor instanceof SpanProcessor) {
-                throw new \TypeError(
-                    sprintf(
-                        'Span Processors should be of type %s, but object of type %s provided',
-                        SpanProcessor::class,
-                        gettype($spanProcessor) == 'object' ? get_class($spanProcessor) : gettype($spanProcessor)
-                    )
-                );
-            }
-        }
-
-        $this->spanProcessors = $spanProcessors;
+        $this->spanProcessors = new MultiSpanProcessor();
     }
 
-    /**
-     * @param SpanProcessor[] $spanProcessors
-     *
-     * @return static
-     */
-    public static function getInstance(array $spanProcessors = []): self
+    public function getTracer(string $name, ?string $version = ''): API\Tracer
     {
-        if (self::$instance instanceof TracerProvider) {
-            return self::$instance;
+        if (isset($this->tracers[$name]) && $this->tracers[$name] instanceof API\Tracer) {
+            return $this->tracers[$name];
         }
 
-        $instance = new TracerProvider($spanProcessors);
+        $spanContext = SpanContext::generate();
 
-        return self::$instance = $instance;
+        return $this->tracers[$name] = new Tracer($this, $spanContext);
     }
 
-    public function getTracer(string $name, ?string $version = ''): API\Tracer
+    public function addSpanProcessor(SpanProcessor $processor): self
     {
-        if (isset($this->tracers[$name]) && $this->tracers[$name] instanceof API\Tracer) {
-            return $this->tracers[$name];
-        }
+        $this->spanProcessors->addSpanProcessor($processor);
 
-        $spanContext = SpanContext::generate();
+        return $this;
+    }
 
-        return $this->tracers[$name] = new Tracer($this->spanProcessors, $spanContext);
+    public function getSpanProcessor(): MultiSpanProcessor
+    {
+        return $this->spanProcessors;
```

Comment:
This is super petty but actually made me re-read things three times to make sure I was following...
The cardinality of the vars and funcs don't line up here and I think it's partly due to holdover naming from before your PR.

Previously there was a class level var that held an array of `SpanProcessors` aptly named `$spanProcessors`. In your new implementation, the var has changed to be a single instance of the meta class `MultiSpanProcessor` (or `SpanProcessorCollection` or whatever it ends up being called) but the var name was unchanged from `$spanProcessors`.

Then we have `getSpanProcessor()` (singular) and returns `$spanProcessors` (plural).

All that to suggest: If we renamed `MultiSpanProcessor` to `SpanProcessorCollector` then the class level var here could be renamed from `$spanProcessors` to `$spanProcessorCollector` and `getSpanProcessor()` could either become `getSpanProcessorCollector()` or at a minimum `getSpanProcessors()` (plural)